### ISSUE TITLE
Regarding unbalanced class distributions in indoor semantic segmentation datasets

### ISSUE THREAD

**Issue by TSchattschneider** [yangyanli/PointCNN#129]
Hello,

I am asking myself to what extend you have analyzed the impact of the class imbalance present in many indoor room datasets. Take the ScanNet dataset for example (which I am especially interested in) and [display the class distribution](https://gist.github.com/TSchattschneider/5794a8c5d5898e16dfde1100e5bffea7):

```
Label          : Percentage of total number of points
------------------------------------------------------
wall           : 36.5%
floor          : 24.4%
unknown        : 11.%
cabinet        : 5.4%
otherfurniture : 3.2%
bed            : 2.9%
door           : 2.4%
bathtub        : 2.4%
curtain        : 2.2%
table          : 2.2%
sofa           : 2.1%
chair          : 1.6%
counter        : 1.%
refridgerator  : 0.7%
desk           : 0.7%
toilet         : 0.3%
window         : 0.3%
picture        : 0.3%
bookshelf      : 0.2%
sink           : 0.2%
shower         : 0.1%
```

As you can see, wall and floor account for over 60% of the points in the dataset, while only about 0.2% of the points belong to objects labeled as sink.

I have noticed that while you provide a method to specifically tune weights for each label, you set them all to the same constant value in your settings ([scannet_x8_2048_fps.py](https://github.com/yangyanli/PointCNN/blob/master/pointcnn_seg/scannet_x8_2048_fps.py#L14), for example). This means that your cross-entropy loss with softmax is unweighted, potentially shifting the model's predictions towards the most common classes.

In your paper, you only provide the overall accuracy as a metric for the segmentation task on ScanNet which could be a problematic metric in the light of this unbalanced distribution. Just getting the most dominant classes right and neglecting the less common classes would suffice for the model to reach a relatively high accuracy. But interestingly, when looking at visualizations of segmentation results of ScanNet, your model nevertheless seems to do a good job e. g. in identifying the chairs, even though they only make up 1.6% of the total points. So this could be less of a problem than it might seem.

Still, table 3 of your paper's supplementary material seems to support my suspicions:

![s3dis-results](https://user-images.githubusercontent.com/7989606/50335684-67c1a080-050c-11e9-8e52-eb2332dd20d4.png)

Here you can see that for the similar S3DIS dataset, most models tend to just get simple structures like the ceiling, floor and wall right, which should account for most of the samples present in the dataset. This again points to the unweighted cross-entropy loss with softmax being problematic when dealing with unbalanced class distributions in the training data.

So my question is, have you experimented with adjusting the label weights and did you see any improvements when trying to compensate for the class imbalance? I guess that you did try to use the weighting, because you have implemented it. Did you have any interesting insights?

**Comment by yangyanli** [comment]
Very good analysis!
Indeed, we tried to weight the loss for compensating the imbalance, which improves mean_per_class_accuracy, while hurting the overall accuracy, but not too much.

We were also surprised that the difference is not that big. The reason why the imbalance is not causing a big problem as expected is not clear.

**Comment by TSchattschneider** [comment]
That's interesting. Getting an improved mean per-class accuracy at the cost of overall accuracy could indicate that the weighting indeed "forces" the model in some sense to get better at segmentation when dealing with less common classes, while at the same time it is *probably* getting a little worse results of the walls / ceilings / floors.

For better insights in these regards, it would be helpful to look at specific accuracies per class and to compare how they change with and without the per-class-weighting. I will further look into this.
